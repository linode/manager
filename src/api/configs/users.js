import {
  genConfig, ReducerGenerator, genActions,
  ONE, MANY, DELETE, PUT, POST,
} from '~/api/apiResultActionReducerGenerator';

export const config = genConfig({
  plural: 'users',
  singular: 'user',
  primaryKey: 'username',
  endpoint: id => `/account/users/${id}`,
  supports: [ONE, MANY, DELETE, PUT, POST],
});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
