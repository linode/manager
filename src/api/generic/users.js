import {
  addParentRefs, ReducerGenerator, genActions,
  ONE, MANY, DELETE, PUT, POST,
} from '~/api/internal.ts';

export const config = addParentRefs({
  name: 'users',
  primaryKey: 'username',
  endpoint: user => `/account/users/${user}`,
  supports: [ONE, MANY, DELETE, PUT, POST],
  subresources: {
    _permissions: {
      name: 'permissions',
      primaryKey: 'id',
      endpoint: user => `/account/users/${user}/grants`,
      supports: [ONE, PUT],
    },
    _password: {
      name: 'password',
      primaryKey: 'id',
      endpoint: user => `/account/users/${user}/password`,
      supports: [POST],
    },
  },
});

export const actions = genActions(config);
export const { reducer } = new ReducerGenerator(config);
