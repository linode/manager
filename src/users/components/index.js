export { default as UserForm } from './UserForm';
