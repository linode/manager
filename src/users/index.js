import React from 'react';
import { Route, IndexRoute } from 'react-router';

import IndexPage from './layouts/IndexPage';

export default (
  <Route path="/users">
    <IndexRoute component={IndexPage} />
  </Route>
);
