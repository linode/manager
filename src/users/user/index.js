import React from 'react';
import { Route, IndexRoute } from 'react-router';

import IndexPage from './layouts/IndexPage';
import EditUserPage from './layouts/EditUserPage';
import PermissionsPage from './layouts/PermissionsPage';

export default (
  <Route path=":username" component={IndexPage}>
    <IndexRoute component={EditUserPage} />
    <Route path="permissions" component={PermissionsPage} />
  </Route>
);
