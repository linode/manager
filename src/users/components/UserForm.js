import React, { Component, PropTypes } from 'react';
import { push } from 'react-router-redux';

import {
  Form, FormGroup, FormGroupError, PasswordInput, Input, Checkboxes, Radio, SubmitButton,
} from 'linode-components/forms';

import { users } from '~/api';
import { setTitle } from '~/actions/title';
import { actions } from '~/api/configs/users';
import { dispatchOrStoreErrors, FormSummary } from '~/components/forms';


export default class UserForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      username: props.user.username,
      email: props.user.email,
      restricted: props.user.restricted,
      password: '',
      loading: false,
      errors: {},
    };
  }

  onChange = ({ target: { name, value, type } }) =>
    this.setState({ [name]: type === 'radio' ? value === 'true' : value })

  onSubmit = () => {
    const { dispatch, user: { username: oldUsername } } = this.props;
    const data = {
      username: this.state.username,
      email: this.state.email,
      restricted: this.state.restricted,
      password: this.state.password,
    };

    if (oldUsername) {
      delete data.password;
    }

    const creating = !oldUsername;

    const idsPath = [oldUsername].filter(Boolean);
    return dispatch(dispatchOrStoreErrors.call(this, [
      () => users[creating ? 'post' : 'put'](data, ...idsPath),
      (user) => creating ? null : actions.one(user),
      () => oldUsername !== data.username && push(`/users/${data.username}`),
      () => oldUsername !== data.username && actions.delete(oldUsername),
      () => creating ? null : setTitle(data.username),
    ]));
  }

  render() {
    const { username, email, restricted, password, loading, errors } = this.state;

    return (
      <Form onSubmit={this.onSubmit}>
        <FormGroup errors={errors} name="username" className="row">
          <label htmlFor="username" className="col-sm-2 col-form-label">Username</label>
          <div className="col-sm-10 clearfix">
            <Input
              name="username"
              id="username"
              value={username}
              onChange={this.onChange}
              className="float-sm-left"
            />
            <FormGroupError errors={errors} name="username" className="float-sm-left" />
          </div>
        </FormGroup>
        <FormGroup errors={errors} name="email" className="row">
          <label htmlFor="email" className="col-sm-2 col-form-label">Email</label>
          <div className="col-sm-10 clearfix">
            <Input
              name="email"
              id="email"
              value={email}
              onChange={this.onChange}
              className="float-sm-left"
            />
            <FormGroupError errors={errors} name="email" className="float-sm-left" />
          </div>
        </FormGroup>
        {this.props.user.username ? null :
          <FormGroup errors={errors} name="password" className="row">
            <label htmlFor="password" className="col-sm-2 col-form-label">Password</label>
            <div className="col-sm-10 clearfix">
              <PasswordInput
                name="password"
                id="password"
                value={password}
                onChange={this.onChange}
              />
              <FormGroupError errors={errors} name="password" className="float-sm-left" />
            </div>
          </FormGroup>
        }
        <FormGroup errors={errors} name="restricted" className="row">
          <label className="col-sm-2 col-form-label">Restricted</label>
          <div className="col-sm-10 clearfix">
            <Checkboxes className="float-sm-left">
              <Radio
                id="restricted"
                name="restricted"
                value
                checked={restricted}
                onChange={this.onChange}
                label="Yes - this user can only do what I specify"
              />
              <Radio
                id="unrestricted"
                name="restricted"
                value={false}
                checked={!restricted}
                onChange={this.onChange}
                label="No - this user has no access restrictions"
              />
            </Checkboxes>
            <FormGroupError errors={errors} name="restricted" className="float-sm-left" />
          </div>
        </FormGroup>
        <div className="row">
          <div className="offset-sm-2 col-sm-10">
            <SubmitButton
              disabled={loading}
              disabledChildren={this.props.user.username ? undefined : 'Adding User'}
            >{this.props.user.username ? undefined : 'Add User'}</SubmitButton>
            <FormSummary
              errors={errors}
              success={this.props.user.username && 'User settings saved.'}
            />
          </div>
        </div>
      </Form>
    );
  }
}

UserForm.propTypes = {
  user: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
};

UserForm.defaultProps = {
  user: {
    username: '',
    email: '',
    restricted: true,
  },
};
