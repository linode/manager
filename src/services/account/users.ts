import { API_ROOT } from 'src/constants';
import Request, { setData, setMethod, setParams, setURL, setXFilter } from 'src/services';

type Page<T> = Linode.ResourcePage<T>;

export const getUsers = (params: any = {}, filters: any = {}) =>
  Request<Page<Linode.User>>(
    setURL(`${API_ROOT}/account/users`),
    setMethod('GET'),
    setParams(params),
    setXFilter(filters),
  )
    .then(response => response.data);

export const getUser = (username: string) =>
  Request<Linode.User>(
    setURL(`${API_ROOT}/account/users/${username}`),
    setMethod('GET'),
  )
    .then(response => response.data);

export const createUser = (data: Partial<Linode.User>) =>
  Request<Linode.User>(
    setURL(`${API_ROOT}/account/users`),
    setMethod('POST'),
    setData(data),
  )
    .then(response => response.data);

export const updateUser = (username: string, data: Partial<Linode.User>) =>
  Request<Linode.User>(
    setURL(`${API_ROOT}/account/users/${username}`),
    setMethod('PUT'),
    setData(data),
  )
    .then(response => response.data);

export const deleteUser = (username: string) =>
  Request<{}>(
    setURL(`${API_ROOT}/account/users/${username}`),
    setMethod('DELETE'),
  )
    .then(response => response.data);

export const getGrants = (username: string) =>
  Request<Linode.Grants>(
    setURL(`${API_ROOT}/account/users/${username}/grants`),
    setMethod('GET'),
  )
    .then(response => response.data);

export const updateGrants = (username: string, data: Partial<Linode.Grants>) =>
  Request<Linode.Grants>(
    setURL(`${API_ROOT}/account/users/${username}/grants`),
    setMethod('PUT'),
    setData(data),
  )
    .then(response => response.data);
