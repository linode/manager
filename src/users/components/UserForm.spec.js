import { shallow } from 'enzyme';
import React from 'react';
import sinon from 'sinon';

import UserForm from '~/users/components/UserForm';

import { testUser } from '~/data/users';


describe('users/layouts/UserForm', () => {
  const sandbox = sinon.sandbox.create();

  afterEach(() => {
    sandbox.restore();
  });

  const dispatch = sandbox.stub();

  it('renders data in UserForm', () => {
    const page = shallow(
      <UserForm
        dispatch={dispatch}
        user={testUser}
      />
    );

    const username = page.find('#username');
    expect(username.props().value).toBe(testUser.username);
    const email = page.find('#email');
    expect(email.props().value).toBe(testUser.email);
  });
});
