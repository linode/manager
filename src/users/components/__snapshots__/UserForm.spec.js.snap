// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users/layouts/UserForm should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserForm
    dispatch={[MockFunction]}
    user={
      Object {
        "_polling": false,
        "email": "example1@domain.com",
        "password": "crisprcas9",
        "restricted": false,
        "username": "testuser1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "analytics": Object {
        "action": "edit",
        "title": "Add/Edit User",
      },
      "children": Array [
        <FormGroup
          className="row"
          errors={Object {}}
          name="username"
        >
          <label
            className="col-sm-2 col-form-label"
            htmlFor="username"
          >
            Username
          </label>
          <div
            className="col-sm-10 clearfix"
          >
            <Input
              className="float-sm-left"
              id="username"
              name="username"
              onChange={[Function]}
              value="testuser1"
            />
            <FormGroupError
              className="float-sm-left"
              errors={Object {}}
              inline={true}
              name="username"
            />
          </div>
        </FormGroup>,
        <FormGroup
          className="row"
          errors={Object {}}
          name="email"
        >
          <label
            className="col-sm-2 col-form-label"
            htmlFor="email"
          >
            Email
          </label>
          <div
            className="col-sm-10 clearfix"
          >
            <Input
              className="float-sm-left"
              id="email"
              name="email"
              onChange={[Function]}
              value="example1@domain.com"
            />
            <FormGroupError
              className="float-sm-left"
              errors={Object {}}
              inline={true}
              name="email"
            />
          </div>
        </FormGroup>,
        null,
        <FormGroup
          className="row"
          errors={Object {}}
          name="restricted"
        >
          <label
            className="col-sm-2 col-form-label"
          >
            Restricted
          </label>
          <div
            className="col-sm-10 clearfix"
          >
            <Checkboxes
              className="float-sm-left"
            >
              <Radio
                checked={false}
                className=""
                id="restricted"
                label="Yes - this user can only do what I specify"
                name="restricted"
                onChange={[Function]}
                value="yes"
              />
              <Radio
                checked={true}
                className=""
                id="unrestricted"
                label="No - this user has no access restrictions"
                name="restricted"
                onChange={[Function]}
                value="no"
              />
            </Checkboxes>
            <FormGroupError
              className="float-sm-left"
              errors={Object {}}
              inline={true}
              name="restricted"
            />
          </div>
        </FormGroup>,
        <div
          className="row"
        >
          <div
            className="offset-sm-2 col-sm-10"
          >
            <SubmitButton
              className="btn-default"
              disabled={false}
              disabledChildren="Saving"
            >
              Save
            </SubmitButton>
            <FormSummary
              className=""
              errors={Object {}}
              success="User settings saved."
            />
          </div>
        </div>,
      ],
      "className": "",
      "noSubmitEvent": false,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <label
              className="col-sm-2 col-form-label"
              htmlFor="username"
            >
              Username
            </label>,
            <div
              className="col-sm-10 clearfix"
            >
              <Input
                className="float-sm-left"
                id="username"
                name="username"
                onChange={[Function]}
                value="testuser1"
              />
              <FormGroupError
                className="float-sm-left"
                errors={Object {}}
                inline={true}
                name="username"
              />
            </div>,
          ],
          "className": "row",
          "errors": Object {},
          "name": "username",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Username",
              "className": "col-sm-2 col-form-label",
              "htmlFor": "username",
            },
            "ref": null,
            "rendered": "Username",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Input
                  className="float-sm-left"
                  id="username"
                  name="username"
                  onChange={[Function]}
                  value="testuser1"
                />,
                <FormGroupError
                  className="float-sm-left"
                  errors={Object {}}
                  inline={true}
                  name="username"
                />,
              ],
              "className": "col-sm-10 clearfix",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "float-sm-left",
                  "id": "username",
                  "name": "username",
                  "onChange": [Function],
                  "value": "testuser1",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "float-sm-left",
                  "errors": Object {},
                  "inline": true,
                  "name": "username",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <label
              className="col-sm-2 col-form-label"
              htmlFor="email"
            >
              Email
            </label>,
            <div
              className="col-sm-10 clearfix"
            >
              <Input
                className="float-sm-left"
                id="email"
                name="email"
                onChange={[Function]}
                value="example1@domain.com"
              />
              <FormGroupError
                className="float-sm-left"
                errors={Object {}}
                inline={true}
                name="email"
              />
            </div>,
          ],
          "className": "row",
          "errors": Object {},
          "name": "email",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Email",
              "className": "col-sm-2 col-form-label",
              "htmlFor": "email",
            },
            "ref": null,
            "rendered": "Email",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Input
                  className="float-sm-left"
                  id="email"
                  name="email"
                  onChange={[Function]}
                  value="example1@domain.com"
                />,
                <FormGroupError
                  className="float-sm-left"
                  errors={Object {}}
                  inline={true}
                  name="email"
                />,
              ],
              "className": "col-sm-10 clearfix",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "float-sm-left",
                  "id": "email",
                  "name": "email",
                  "onChange": [Function],
                  "value": "example1@domain.com",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "float-sm-left",
                  "errors": Object {},
                  "inline": true,
                  "name": "email",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": [Function],
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <label
              className="col-sm-2 col-form-label"
            >
              Restricted
            </label>,
            <div
              className="col-sm-10 clearfix"
            >
              <Checkboxes
                className="float-sm-left"
              >
                <Radio
                  checked={false}
                  className=""
                  id="restricted"
                  label="Yes - this user can only do what I specify"
                  name="restricted"
                  onChange={[Function]}
                  value="yes"
                />
                <Radio
                  checked={true}
                  className=""
                  id="unrestricted"
                  label="No - this user has no access restrictions"
                  name="restricted"
                  onChange={[Function]}
                  value="no"
                />
              </Checkboxes>
              <FormGroupError
                className="float-sm-left"
                errors={Object {}}
                inline={true}
                name="restricted"
              />
            </div>,
          ],
          "className": "row",
          "errors": Object {},
          "name": "restricted",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Restricted",
              "className": "col-sm-2 col-form-label",
            },
            "ref": null,
            "rendered": "Restricted",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Checkboxes
                  className="float-sm-left"
                >
                  <Radio
                    checked={false}
                    className=""
                    id="restricted"
                    label="Yes - this user can only do what I specify"
                    name="restricted"
                    onChange={[Function]}
                    value="yes"
                  />
                  <Radio
                    checked={true}
                    className=""
                    id="unrestricted"
                    label="No - this user has no access restrictions"
                    name="restricted"
                    onChange={[Function]}
                    value="no"
                  />
                </Checkboxes>,
                <FormGroupError
                  className="float-sm-left"
                  errors={Object {}}
                  inline={true}
                  name="restricted"
                />,
              ],
              "className": "col-sm-10 clearfix",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    <Radio
                      checked={false}
                      className=""
                      id="restricted"
                      label="Yes - this user can only do what I specify"
                      name="restricted"
                      onChange={[Function]}
                      value="yes"
                    />,
                    <Radio
                      checked={true}
                      className=""
                      id="unrestricted"
                      label="No - this user has no access restrictions"
                      name="restricted"
                      onChange={[Function]}
                      value="no"
                    />,
                  ],
                  "className": "float-sm-left",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "checked": false,
                      "className": "",
                      "id": "restricted",
                      "label": "Yes - this user can only do what I specify",
                      "name": "restricted",
                      "onChange": [Function],
                      "value": "yes",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "checked": true,
                      "className": "",
                      "id": "unrestricted",
                      "label": "No - this user has no access restrictions",
                      "name": "restricted",
                      "onChange": [Function],
                      "value": "no",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "float-sm-left",
                  "errors": Object {},
                  "inline": true,
                  "name": "restricted",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="offset-sm-2 col-sm-10"
          >
            <SubmitButton
              className="btn-default"
              disabled={false}
              disabledChildren="Saving"
            >
              Save
            </SubmitButton>
            <FormSummary
              className=""
              errors={Object {}}
              success="User settings saved."
            />
          </div>,
          "className": "row",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SubmitButton
                className="btn-default"
                disabled={false}
                disabledChildren="Saving"
              >
                Save
              </SubmitButton>,
              <FormSummary
                className=""
                errors={Object {}}
                success="User settings saved."
              />,
            ],
            "className": "offset-sm-2 col-sm-10",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Save",
                "className": "btn-default",
                "disabled": false,
                "disabledChildren": "Saving",
              },
              "ref": null,
              "rendered": "Save",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "",
                "errors": Object {},
                "success": "User settings saved.",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "analytics": Object {
          "action": "edit",
          "title": "Add/Edit User",
        },
        "children": Array [
          <FormGroup
            className="row"
            errors={Object {}}
            name="username"
          >
            <label
              className="col-sm-2 col-form-label"
              htmlFor="username"
            >
              Username
            </label>
            <div
              className="col-sm-10 clearfix"
            >
              <Input
                className="float-sm-left"
                id="username"
                name="username"
                onChange={[Function]}
                value="testuser1"
              />
              <FormGroupError
                className="float-sm-left"
                errors={Object {}}
                inline={true}
                name="username"
              />
            </div>
          </FormGroup>,
          <FormGroup
            className="row"
            errors={Object {}}
            name="email"
          >
            <label
              className="col-sm-2 col-form-label"
              htmlFor="email"
            >
              Email
            </label>
            <div
              className="col-sm-10 clearfix"
            >
              <Input
                className="float-sm-left"
                id="email"
                name="email"
                onChange={[Function]}
                value="example1@domain.com"
              />
              <FormGroupError
                className="float-sm-left"
                errors={Object {}}
                inline={true}
                name="email"
              />
            </div>
          </FormGroup>,
          null,
          <FormGroup
            className="row"
            errors={Object {}}
            name="restricted"
          >
            <label
              className="col-sm-2 col-form-label"
            >
              Restricted
            </label>
            <div
              className="col-sm-10 clearfix"
            >
              <Checkboxes
                className="float-sm-left"
              >
                <Radio
                  checked={false}
                  className=""
                  id="restricted"
                  label="Yes - this user can only do what I specify"
                  name="restricted"
                  onChange={[Function]}
                  value="yes"
                />
                <Radio
                  checked={true}
                  className=""
                  id="unrestricted"
                  label="No - this user has no access restrictions"
                  name="restricted"
                  onChange={[Function]}
                  value="no"
                />
              </Checkboxes>
              <FormGroupError
                className="float-sm-left"
                errors={Object {}}
                inline={true}
                name="restricted"
              />
            </div>
          </FormGroup>,
          <div
            className="row"
          >
            <div
              className="offset-sm-2 col-sm-10"
            >
              <SubmitButton
                className="btn-default"
                disabled={false}
                disabledChildren="Saving"
              >
                Save
              </SubmitButton>
              <FormSummary
                className=""
                errors={Object {}}
                success="User settings saved."
              />
            </div>
          </div>,
        ],
        "className": "",
        "noSubmitEvent": false,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <label
                className="col-sm-2 col-form-label"
                htmlFor="username"
              >
                Username
              </label>,
              <div
                className="col-sm-10 clearfix"
              >
                <Input
                  className="float-sm-left"
                  id="username"
                  name="username"
                  onChange={[Function]}
                  value="testuser1"
                />
                <FormGroupError
                  className="float-sm-left"
                  errors={Object {}}
                  inline={true}
                  name="username"
                />
              </div>,
            ],
            "className": "row",
            "errors": Object {},
            "name": "username",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Username",
                "className": "col-sm-2 col-form-label",
                "htmlFor": "username",
              },
              "ref": null,
              "rendered": "Username",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Input
                    className="float-sm-left"
                    id="username"
                    name="username"
                    onChange={[Function]}
                    value="testuser1"
                  />,
                  <FormGroupError
                    className="float-sm-left"
                    errors={Object {}}
                    inline={true}
                    name="username"
                  />,
                ],
                "className": "col-sm-10 clearfix",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "float-sm-left",
                    "id": "username",
                    "name": "username",
                    "onChange": [Function],
                    "value": "testuser1",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "float-sm-left",
                    "errors": Object {},
                    "inline": true,
                    "name": "username",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <label
                className="col-sm-2 col-form-label"
                htmlFor="email"
              >
                Email
              </label>,
              <div
                className="col-sm-10 clearfix"
              >
                <Input
                  className="float-sm-left"
                  id="email"
                  name="email"
                  onChange={[Function]}
                  value="example1@domain.com"
                />
                <FormGroupError
                  className="float-sm-left"
                  errors={Object {}}
                  inline={true}
                  name="email"
                />
              </div>,
            ],
            "className": "row",
            "errors": Object {},
            "name": "email",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Email",
                "className": "col-sm-2 col-form-label",
                "htmlFor": "email",
              },
              "ref": null,
              "rendered": "Email",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Input
                    className="float-sm-left"
                    id="email"
                    name="email"
                    onChange={[Function]}
                    value="example1@domain.com"
                  />,
                  <FormGroupError
                    className="float-sm-left"
                    errors={Object {}}
                    inline={true}
                    name="email"
                  />,
                ],
                "className": "col-sm-10 clearfix",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "float-sm-left",
                    "id": "email",
                    "name": "email",
                    "onChange": [Function],
                    "value": "example1@domain.com",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "float-sm-left",
                    "errors": Object {},
                    "inline": true,
                    "name": "email",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <label
                className="col-sm-2 col-form-label"
              >
                Restricted
              </label>,
              <div
                className="col-sm-10 clearfix"
              >
                <Checkboxes
                  className="float-sm-left"
                >
                  <Radio
                    checked={false}
                    className=""
                    id="restricted"
                    label="Yes - this user can only do what I specify"
                    name="restricted"
                    onChange={[Function]}
                    value="yes"
                  />
                  <Radio
                    checked={true}
                    className=""
                    id="unrestricted"
                    label="No - this user has no access restrictions"
                    name="restricted"
                    onChange={[Function]}
                    value="no"
                  />
                </Checkboxes>
                <FormGroupError
                  className="float-sm-left"
                  errors={Object {}}
                  inline={true}
                  name="restricted"
                />
              </div>,
            ],
            "className": "row",
            "errors": Object {},
            "name": "restricted",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Restricted",
                "className": "col-sm-2 col-form-label",
              },
              "ref": null,
              "rendered": "Restricted",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Checkboxes
                    className="float-sm-left"
                  >
                    <Radio
                      checked={false}
                      className=""
                      id="restricted"
                      label="Yes - this user can only do what I specify"
                      name="restricted"
                      onChange={[Function]}
                      value="yes"
                    />
                    <Radio
                      checked={true}
                      className=""
                      id="unrestricted"
                      label="No - this user has no access restrictions"
                      name="restricted"
                      onChange={[Function]}
                      value="no"
                    />
                  </Checkboxes>,
                  <FormGroupError
                    className="float-sm-left"
                    errors={Object {}}
                    inline={true}
                    name="restricted"
                  />,
                ],
                "className": "col-sm-10 clearfix",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      <Radio
                        checked={false}
                        className=""
                        id="restricted"
                        label="Yes - this user can only do what I specify"
                        name="restricted"
                        onChange={[Function]}
                        value="yes"
                      />,
                      <Radio
                        checked={true}
                        className=""
                        id="unrestricted"
                        label="No - this user has no access restrictions"
                        name="restricted"
                        onChange={[Function]}
                        value="no"
                      />,
                    ],
                    "className": "float-sm-left",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "checked": false,
                        "className": "",
                        "id": "restricted",
                        "label": "Yes - this user can only do what I specify",
                        "name": "restricted",
                        "onChange": [Function],
                        "value": "yes",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "checked": true,
                        "className": "",
                        "id": "unrestricted",
                        "label": "No - this user has no access restrictions",
                        "name": "restricted",
                        "onChange": [Function],
                        "value": "no",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                  ],
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "float-sm-left",
                    "errors": Object {},
                    "inline": true,
                    "name": "restricted",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="offset-sm-2 col-sm-10"
            >
              <SubmitButton
                className="btn-default"
                disabled={false}
                disabledChildren="Saving"
              >
                Save
              </SubmitButton>
              <FormSummary
                className=""
                errors={Object {}}
                success="User settings saved."
              />
            </div>,
            "className": "row",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <SubmitButton
                  className="btn-default"
                  disabled={false}
                  disabledChildren="Saving"
                >
                  Save
                </SubmitButton>,
                <FormSummary
                  className=""
                  errors={Object {}}
                  success="User settings saved."
                />,
              ],
              "className": "offset-sm-2 col-sm-10",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "Save",
                  "className": "btn-default",
                  "disabled": false,
                  "disabledChildren": "Saving",
                },
                "ref": null,
                "rendered": "Save",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "",
                  "errors": Object {},
                  "success": "User settings saved.",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
